GPU matching algorithm implementation

majabojarska opened the pull request:
Each kernel performs pattern matching on 128 positions.

plonajakub committed:
GPU implementation - WIP

majabojarska committed:
Working GPU implementation.

Co-authored-by: Jakub Plona <plonajakub@outlook.com>

majabojarska commented on webpack.config.ts:
```diff
@@ -85,12 +85,12 @@ export default {
+      //     // fullySpecified: false,
+      //   },
+      // },
     ],
```
Before commenting out this was causing crashes, on a clean environment setup. @damiankoper  please take a look.

damiankoper replied:
The purpose of this was to keep benchmark.js sane. With this lib gone, there is no need to keep this setting. Can be removed.

majabojarska replied:
Removed this piece of code.

damiankoper commented on src/domain/gpu/Payloads.ts:
```diff
@@ -0,0 +1,9 @@
+export interface DataInMessagePayload {
+  pattern: Uint8Array;
+  text: Uint8Array;
+}
+
+export interface DataOutMessagePayload {
+  matches: Array<number>;
+  error: Error;
+}
```
I will refactor this after merge alongside with proper scheduler with shared interface.

damiankoper commented on src/domain/gpu/Solver.ts:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  /**
+   * Each kernel handles 128 positions.
+   * Returns Array<number> of size 4.
```
Is it worth to have this configurable?
In CPU we can set number of workers `[1; inf]`.
So here we can set size of data processed by single kernel `[1; 4]`.

damiankoper replied:
I suggest you/we test this for variety of cases since I'm not sure about 32b int representation.
Does `console.log` log instance of `Uint32Array`?

majabojarska replied:
We could, but there is already quite a bit of information we will get from the GPU vs CPU implementations. Benchmarking and analysis of that many cases will be time-consuming.

majabojarska replied:
What do you think @plonajakub ?

majabojarska replied:
Regarding data types, maybe we should look into the following:
```typescript
kernel.setPrecision('unsigned' | 'single');
```

plonajakub replied:
Yeah, it can be only seen in the performance tests if there is a need of having this configurable. I would discuss it face to face.

majabojarska replied:
We decided with @plonajakub to make kernel coverage configurable to 1,2,4 32-bit words (32,64,128 exact matching positions). Should be done by November 4th.

damiankoper reviewed (changes requested):
Not really changes but we have to make sure this method works. Check my comments.

majabojarska committed:
Remove obsolete entry in webpack config

majabojarska committed:
Remove ASCII art

majabojarska committed:
Add mode parameter to SolverGPU

damiankoper committed:
refactor: performance & style

damiankoper committed:
feat: GpuScheduler & interface refactor

plonajakub commented on src/domain/gpu/Solver.ts:
```diff
@@ -0,0 +1,144 @@
+    const startPosition: number = this.thread.x * 128;
+
+    // Allocate 32x4 bits, each bit is a match flag.
+    const matches: number[] = [0, 0, 0, 0];
```
Could we extract it and consider it as a parameter? That means setting the buffer size outside the kernel function.

plonajakub replied:
A similar topic is discussed above so I am resolving this.

plonajakub reviewed (approved):
Really good job! There are still some places for efficiency improvement, however, these changes make a solid base.

majabojarska committed:
Remove obsolete code.

Co-authored-by: Jakub Plona <plonajakub@outlook.com>

majabojarska committed:
Save work in progress.

- Fixed fallback to CPU - Algo runs on GPU now.
- Words 32b per kernel parametrization
- Match position extraction optimization

Co-authored-by: Jakub Plona <plonajakub@outlook.com>

majabojarska committed:
Refactor SolverGPU.extractMatches

damiankoper committed:
Unified scheduler interface
